Clear error if applicationWrapperPath does not exist

Right now if you specify an `applicationWrapperPath` and the file does not exist, you get an error like this:

```
./node_modules/@mapbox/batfish/dist/webpack/batfish-app.js
Module not found: Can't resolve 'batfish-internal/application-wrapper' in 
'/some/path/to/node_modules/@mapbox/batfish/dist/webpack'
```

We should make it more clear what happened. The user does not and should not need to know what `batfish-internal/application-wrapper` is.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,8 @@
 ## Head
 
 - [Fix] Add Prettier to `dependencies`, not just `devDependencies`.
+- [Fix] Provide more clear error messages by checking for the existing of files or directories while validating configuration for the following options: `applicationWrapperPath`, `pagesDirectory`, `outputDirectory`, `temporaryDirectory`, `inlineJs`.
+- [Fix] Fix buggy validation of `inlineJs` configuration property.
 
 ## 0.13.0
 

--- a/examples/miscellany/batfish.config.js
+++ b/examples/miscellany/batfish.config.js
@@ -5,6 +5,10 @@ module.exports = () => {
   return {
     siteBasePath: '/miscellany',
     siteOrigin: 'https://www.your-batfish-site.com',
+    applicationWrapperPath: path.join(
+      __dirname,
+      './src/components/application-wrapper.js'
+    ),
     stylesheets: [
       'https://api.mapbox.com/mapbox-assembly/v0.17.0/assembly.min.css'
     ],

--- a/examples/miscellany/src/components/application-wrapper.js
+++ b/examples/miscellany/src/components/application-wrapper.js
@@ -1,0 +1,7 @@
+import React from 'react';
+
+export default class ApplicationWrapper extends React.Component {
+  render() {
+    return <div id="application-wrapper">{this.props.children}</div>;
+  }
+}

--- a/src/node/validate-config.js
+++ b/src/node/validate-config.js
@@ -6,10 +6,10 @@ const del = require('del');
 const path = require('path');
 const mkdirp = require('mkdirp');
 const chalk = require('chalk');
-const fs = require('fs');
 const a = require('indefinite');
 const isGlob = require('is-glob');
 const isAbsoluteUrl = require('is-absolute-url');
+const pathType = require('path-type');
 const autoprefixer = require('autoprefixer');
 const errorTypes = require('./error-types');
 
@@ -30,8 +30,8 @@ const configSchema = {
     description: 'string'
   },
   applicationWrapperPath: {
-    validator: isAbsolutePath,
-    description: 'absolute path'
+    validator: isAbsolutePathToExistingFile,
+    description: 'absolute path to an existing file'
   },
   stylesheets: {
     validator: isArrayOf(isValidStylesheetItem),
@@ -43,16 +43,16 @@ const configSchema = {
     description: 'string or array of strings'
   },
   pagesDirectory: {
-    validator: isAbsolutePath,
-    description: 'absolute path'
+    validator: isAbsolutePathToExistingDirectory,
+    description: 'absolute path to an existing directory'
   },
   outputDirectory: {
-    validator: isAbsolutePath,
-    description: 'absolute path'
+    validator: isAbsolutePathToExistingDirectory,
+    description: 'absolute path to an existing directory'
   },
   temporaryDirectory: {
-    validator: isAbsolutePath,
-    description: 'absolute path'
+    validator: isAbsolutePathToExistingDirectory,
+    description: 'absolute path to an existing directory'
   },
   dataSelectors: {
     validator: x => _.isPlainObject(x) && _.every(x, _.isFunction),
@@ -120,7 +120,12 @@ const configSchema = {
     description: 'boolean'
   },
   inlineJs: {
-    validator: isArrayOf(x => isAbsolutePath(x.filename) || _.isBoolean(x)),
+    validator: isArrayOf(x => {
+      return (
+        isAbsolutePathToExistingFile(x.filename) &&
+        (x.uglify === undefined || _.isBoolean(x))
+      );
+    }),
     description:
       'array of objects with an absolute path for the "filename" prop, and an optional "boolean" uglify prop'
   },
@@ -330,7 +335,14 @@ function isExistingFile(value: *): boolean {
   if (!_.isString(value)) {
     return false;
   }
-  return fs.existsSync(value);
+  return pathType.fileSync(value);
+}
+
+function isExistingDirectory(value: *): boolean {
+  if (!_.isString(value)) {
+    return false;
+  }
+  return pathType.dirSync(value);
 }
 
 function isValidStylesheetItem(value: *): boolean {
@@ -339,4 +351,12 @@ function isValidStylesheetItem(value: *): boolean {
   }
   if (!_.isString(value)) return false;
   return isAbsoluteUrl(value) || isAbsolutePath(value);
+}
+
+function isAbsolutePathToExistingFile(value: *): boolean {
+  return isAbsolutePath(value) && isExistingFile(value);
+}
+
+function isAbsolutePathToExistingDirectory(value: *): boolean {
+  return isAbsolutePath(value) && isExistingDirectory(value);
 }

--- a/test/examples.test.js
+++ b/test/examples.test.js
@@ -6,6 +6,8 @@ const fs = require('fs');
 const path = require('path');
 const pathType = require('path-type');
 
+jest.unmock('path-type');
+
 const examplesDirectory = path.join(__dirname, '../examples');
 
 describe('examples', () => {

--- a/test/test-util/jest-setup.js
+++ b/test/test-util/jest-setup.js
@@ -5,3 +5,11 @@ const Enzyme = require('enzyme');
 const Adapter = require('enzyme-adapter-react-16');
 
 Enzyme.configure({ adapter: new Adapter() });
+
+// This module makes it harder to mock things for tests — so mock *it*!
+jest.mock('path-type', () => {
+  return {
+    dirSync: jest.fn(() => true),
+    fileSync: jest.fn(() => true)
+  };
+});

--- a/test/validate-config.test.js
+++ b/test/validate-config.test.js
@@ -5,6 +5,7 @@ const del = require('del');
 const path = require('path');
 const fs = require('fs');
 const stripAnsi = require('strip-ansi');
+const pathType = require('path-type');
 const validateConfig = require('../src/node/validate-config');
 const errorTypes = require('../src/node/error-types');
 const projectRootSerializer = require('./test-util/project-root-serializer');
@@ -219,6 +220,9 @@ describe('validateConfig', () => {
 
   test('stylesheets file paths must exist', () => {
     expect.hasAssertions();
+    pathType.fileSync.mockImplementation(input => {
+      return !/does-not-exist/.test(input);
+    });
     try {
       validateConfig({
         stylesheets: [

